#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd

In [2]:
data_path = '../data/supermarket_sales.csv'
df = pd.read_csv(data_path)

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [6]:
branch_gross = df.groupby('Branch')['gross income'].sum()
branch_gross

Branch
A    5057.1605
B    5057.0320
C    5265.1765
Name: gross income, dtype: float64

Найвищий прибуток отримано філіалом "С", прибутки отримані філіалами "А" та "В" мають невелику різницю між собою.

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [10]:
branch_gross1 = pd.pivot_table(df, index=['Branch','City'], values=['gross income'], aggfunc="sum")
branch_gross1 

gross income
Branch City                   
A      Yangon        5057.1605
B      Mandalay      5057.0320
C      Naypyitaw     5265.1765

Так як індексу філіалу супермаркету "С" відповідає місто Naypyitaw і в ньому отримано найвищий прибуток, то доцільно розглянути думку про відкриття ще одного магазину у цьому місті.

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [35]:
invoices_by_category = pd.pivot_table(df, index='Branch' , columns=['Product line'], values=['Invoice ID'], aggfunc='count')
invoices_by_category

Invoice ID                                         \
Product line Electronic accessories Fashion accessories Food and beverages   
Branch                                                                       
A                                60                  51                 58   
B                                55                  62                 50   
C                                55                  65                 66   

                                                                     
Product line Health and beauty Home and lifestyle Sports and travel  
Branch                                                               
A                           47                 65                59  
B                           53                 50                62  
C                           52                 45                45

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [25]:
print(invoices_by_category.columns)

MultiIndex([('Invoice ID', 'Electronic accessories'),
            ('Invoice ID',    'Fashion accessories'),
            ('Invoice ID',     'Food and beverages'),
            ('Invoice ID',      'Health and beauty'),
            ('Invoice ID',     'Home and lifestyle'),
            ('Invoice ID',      'Sports and travel')],
           names=[None, 'Product line'])


In [42]:
branch_with_max = invoices_by_category[('Invoice ID', 'Electronic accessories')].idxmax()

print(f"Філіал із найбільшою кількістю інвойсів у категорії 'Electronic accessories': {branch_with_max}")

Філіал із найбільшою кількістю інвойсів у категорії 'Electronic accessories': A


In [43]:
 invoices_by_category.columns.droplevel(0)

Index(['Electronic accessories', 'Fashion accessories', 'Food and beverages',
       'Health and beauty', 'Home and lifestyle', 'Sports and travel'],
      dtype='object', name='Product line')

In [47]:
invoices_by_category.columns = invoices_by_category.columns.droplevel(0)
invoices_by_category

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45

In [48]:
print(invoices_by_category)

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45  


In [49]:
branch_with_max = invoices_by_category['Electronic accessories'].idxmax()

print(f"Філіал із найбільшою кількістю інвойсів у категорії 'Electronic accessories': {branch_with_max}")

Філіал із найбільшою кількістю інвойсів у категорії 'Electronic accessories': A


4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


Можу запропонувати такі питання:
1. який середній рейтинг задоволення клієнтів від загального досвіду покупок по трьом філіалам супрмаркетів (по шкалі від 1 до 10)
2. який розподіл кількості покупок між кожним типом клієнтів по кожному супермаркету
3. який розподіл використання методів оплати по супермаркетах в розрізі загальної суми покупки та середнього чеку

Розглянемо кожну відповіль окремо.
    1. середній рейтинг задоволення клієнтів від загального досвіду покупок по трьом філіалам супрмаркетів (по шкалі від 1 до 10):

In [57]:
df.groupby('City')['Rating'].mean().round(4)

City
Mandalay     6.8181
Naypyitaw    7.0729
Yangon       7.0271
Name: Rating, dtype: float64

З розрахунку видно, що розбіжності рейтингах між супермаркетах в містах не значні, всі вище середнього значення (шкала вимірювання від 1 до 10), найбільше значення в місті Naypyitaw - 7.0729, найменше в місті Mandalay - 6.8181. 
Керівництву супермаркету в місті Mandalay необхідно направити зусилля на підвищення рейтенгу задоволеності клієнтів через:
- приділення більше уваги підвищенню рівня обслуговування;
- проведення тренінгів для персоналу;
- покращення умов покупок;
- проведення маркетингових кампаній для залучення уваги до цього супермаркету;
- виявлення конкретних проблем або областей для покращення (наприклад, час очікування в черзі, асортимент товарів чи поведінка персоналу).

    2. розподіл кількості покупок між кожним типом клієнтів по кожному супермаркету

In [69]:
invoices_by_ustomer_type = pd.pivot_table(df, index='City' , columns=['Customer type'], values=['Quantity'], aggfunc='sum')
invoices_by_ustomer_type

Quantity       
Customer type   Member Normal
City                         
Mandalay           924    896
Naypyitaw          897    934
Yangon             964    895

In [71]:
customer_type_member_max = invoices_by_ustomer_type[('Quantity', 'Member')].idxmax()
print(f"Місцезнаходження супермаркета із найбільшою кількістю покупок здійсненою клієнтами, що використовують картку члена: {customer_type_member_max}")

Місцезнаходження супермаркета із найбільшою кількістю покупок здійсненою клієнтами, що використовують картку члена: Yangon


In [75]:
customer_type_member_min = invoices_by_ustomer_type[('Quantity', 'Member')].idxmin()
print(f"Місцезнаходження супермаркета із найменшою кількістю покупок здійсненою клієнтами, що використовують картку члена: {customer_type_member_min}")

Місцезнаходження супермаркета із найменшою кількістю покупок здійсненою клієнтами, що використовують картку члена: Naypyitaw


In [77]:
customer_type_normal_max = invoices_by_ustomer_type[('Quantity', 'Normal')].idxmax()
print(f"Місцезнаходження супермаркета із найбільшою кількістю покупок, здійсненою звичайними клієнтами: {customer_type_normal_max}")

Місцезнаходження супермаркета із найбільшою кількістю покупок, здійсненою звичайними клієнтами: Naypyitaw


In [78]:
customer_type_normal_min = invoices_by_ustomer_type[('Quantity', 'Normal')].idxmin()
print(f"Місцезнаходження супермаркета із найменшою кількістю покупок, здійсненою звичайними клієнтами: {customer_type_normal_min}")

Місцезнаходження супермаркета із найменшою кількістю покупок, здійсненою звичайними клієнтами: Yangon


З розрахунку можемо зробити висновок про те, що:
- у місті Yangon знаходиться супермаркет із найбільшою кількістю покупок (964) здійсненою клієнтами, що використовують картку члена та з найменшою кількістю покупок (895), здійсненою звичайними клієнтами;
- у місті Naypyitaw знаходиться супермаркет із найбільшою кількістю покупок (934) здійсненою звичайними клієнтами та з найменшою кількістю покупок (897) здійсненою клієнтами, що використовують картку члена;
- у місті Mandalay середні показники у цих категоріях, при чому кількість кількістю покупок, здійсненою звичайними клієнтами, становить 896, що робить її майже однаковою з показником в місті Yangon;.
Для міста Naypyitaw необхідно зосередитись на просуванні спеціальних акцій, для мотивації  покупців ставати членами програми лояльності.
Для міста Yangon необхідно продовжувати розвивати програму лояльності, пропонуючи нові переваги та знижки, щоб збільшити частоту покупок серед цієї групи, ці захди можна просувати і для міста Mandalay для збільшення кількості покупців, що є членами клубу.   
Моніторинг ефективності програм лояльності в різних містах: Порівняння результатів між різними містами та типами клієнтів допоможе зрозуміти, де програма лояльності дає найбільший ефект, і зосередити зусилля на цих локаціях.

    3. розподіл використання методів оплати по супермаркетах в розрізі загальної суми покупки та середнього чеку

In [55]:
df.groupby(['City', 'Payment']).agg({'Quantity': 'sum', 'Total': 'mean'})

Quantity       Total
City      Payment                          
Mandalay  Cash              628  321.267832
          Credit card       599  342.613362
          Ewallet           593  296.578354
Naypyitaw Cash              696  347.466593
          Credit card       543  309.463929
          Ewallet           592  350.522491
Yangon    Cash              572  307.102282
          Credit card       580  318.218755
          Ewallet           707  312.098167

Проаналізуємо способи оплати по містам:
-  Mandalay - найбільше покупок має спосіб оплати готівкю, найбільша середня покупка оплачена кредитною картою;
-  Naypyitaw - найбільше покупок має спосіб оплати готівкю, найбільша середня покупка оплачена з електронного гаманця;
-  Yangon - найбільше покупок має спосіб оплати електронним гаманцем, найбільша середня покупка оплачена кредитною картою.
Ці дані можна викристовувати для розробки маркетингових компаній для залучення покупців з огляду на спосіб оплати в супермаркетах, пропонуати певні акції з огляду на найбільш популярні способи оплати в кожному місті.